packages/firestore/src/api/reference_impl.ts

```diff
@@ -859,3 +1074,244 @@ function convertToDocSnapshot<AppModelType, DbModelType extends DocumentData>(
+): Unsubscribe {
+  let unsubscribed: boolean = false;
+  let internalUnsubscribe: Unsubscribe | undefined;
+  const loadTask = loadBundle(db, json.bundle);
+  loadTask
+    .then(() => {
+      const docReference = new DocumentReference(
+        db,
+        converter ? converter : null,
+        DocumentKey.fromPath(json.bundleName)
+      );
+      if (options !== undefined) {
+        internalUnsubscribe = onSnapshot(
+          docReference as DocumentReference<AppModelType, DbModelType>,
+          options,
+          observer
+        );
+      } else {
+        internalUnsubscribe = onSnapshot(
+          docReference as DocumentReference<AppModelType, DbModelType>,
+          observer
+        );
+      }
+    })
+    .catch(e => {
+      if (observer.error) {
+        observer.error(e);
+      }
+      return () => {};
+    });
+  return () => {
+    if (unsubscribed) {
+      return;
+    }
+    unsubscribed = true;
+    if (internalUnsubscribe) {
+      internalUnsubscribe();
+    }
+  };
+}
+
+/**
+ * Loads the bundle in a separate task and then invokes {@link onSnapshot} with a
+ * {@link Query} that represents the Query in the bundle.
+ *
+ * @param firestore - The {@link Firestore} instance for the {@link onSnapshot} operation request.
+ * @param json - The JSON bundle to load, produced by {@link QuerySnapshot.toJSON}.
+ * @param options - Options controlling the listen behavior.
+ * @param observer - A single object containing `next` and `error` callbacks.
+ * @param converter - An optional object that converts objects from Firestore before the onNext
+ * listener is invoked.
+ * @returns An unsubscribe function that can be called to cancel the snapshot
+ * listener.
+ *
+ * @internal
+ */
+function onSnapshotQuerySnapshotBundle<
+  AppModelType,
+  DbModelType extends DocumentData
+>(
+  db: Firestore,
+  json: { bundle: string; bundleName: string; bundleSource: string },
+  options: SnapshotListenOptions | undefined,
+  observer: {
+    next: (snapshot: QuerySnapshot<AppModelType, DbModelType>) => void;
+    error?: (error: FirestoreError) => void;
+    complete?: () => void;
+  },
+  converter?: FirestoreDataConverter<DbModelType>
+): Unsubscribe {
+  let unsubscribed: boolean = false;
+  let internalUnsubscribe: Unsubscribe | undefined;
+  const loadTask = loadBundle(db, json.bundle);
+  loadTask
+    .then(() => namedQuery(db, json.bundleName))
+    .then(query => {
+      if (query && !unsubscribed) {
+        const realQuery: Query = (query as Query)!;
+        if (converter) {
+          realQuery.withConverter(converter);
+        }
+        if (options !== undefined) {
+          internalUnsubscribe = onSnapshot(
+            query as Query<AppModelType, DbModelType>,
+            options,
+            observer
+          );
+        } else {
+          internalUnsubscribe = onSnapshot(
+            query as Query<AppModelType, DbModelType>,
+            observer
+          );
```

Comment:
also feels like you could reduce the number of calls to onsnapshot, if there's a default options value that you could pass